### PR TITLE
NonPrefix Thread  Command

### PR DESCRIPTION
**Pull request by lupatrick13**
- 

1.  Able to create threads using an input ID
2. Add message using a thread ID search (Regex)
3.  If there are multiple thread results, user can look at a list of threads and rechoose an ID by typing it in

### CODE REVIEW

**Comment by jerrylee17** [comment]
`npm run lint`

Also rebase patrick branch (it's 2 commits behind master)

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 1]
```diff
@@ -1,17 +1,298 @@
...
+    );
+  };
+  const emojiCollector = currentMsg.createReactionCollector(filter);
+  const checkEmoji = async reaction => {
+    let change = true;      
+    /**
+     *  Handles what the current page is and 
+     *  what the range of threads we want is
+     **/  
+    switch (reaction.emoji.name) {
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/threads/threadMessage.js:204]: Wrapping around would be nice (like set the page to 0 by going right from the last page)

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 2]
```diff
@@ -1,17 +1,298 @@
...
+    return (
+      /** 
+       * if regex matches (new id starts with old id) and 
+       * authors are same then message is collected
+       **/ 
+      newID != null &&    
+      m.author.id === message.author.id
+    );
+  };
+  const messageCollector = message.channel.createMessageCollector(filter);
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/threads/threadMessage.js:269]: Add a time limit to the collectors (end collection and delete displayed message after like a few minutes)

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 3]
```diff
@@ -1,17 +1,298 @@
...
+ * IMPORTANT: HAS TO BE RESPONSE FROM FIND_THREAD
+ **/
+async function multipleThreadResults(data, message, queryThread){
+  /**
+     * @var {Enumeration} page This var contains all necessary 
+     * data to create a "book" of data
+     * @var {array} threadListEmbed this array holds each page.
+     * Each page is a RichEmbed
+     **/
+  let page = {
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/threads/threadMessage.js:159]: If all the threads fit on one page, there's no need for pagination.

**Review by charlotte-zhuang — CHANGES_REQUESTED** [review]
I made a few comments.

1. Documentation comment changes
2. Add example field to command
3. Don't paginate threads if they fit on one page
3. Wrap around on paginated threads
4. Stop waiting for user response after a certain amount of time

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 4]
```diff
@@ -1,17 +1,314 @@
...
+    currentPage: 0,
+    maxPage: Math.ceil(queryThread.responseData.length / ITEM_PER_PAGE) - 1
+  };
+  let threadListEmbed = new Array(page.maxPage);
+  threadListEmbed[0] = new Discord.RichEmbed()
+    .setColor('#301934')
+    .setTitle(`All threads that start with ID: ${data.threadID}`);
+  if(messageMode)
+  {
+    threadListEmbed[0].addField('Choose one!', 'Example: "1"');
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/threads/threadMessage.js:226]: I think this field and the blank field look really weird. What do you think of changing it to `threadListEmbed[0].setDescription('Choose one! Example: "1"')`? Same thing on line 287.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 5]
```diff
@@ -4,22 +4,33 @@ const Command = require('../Command');
...
 module.exports = new Command({
   name: 'no-prefix threadmessage',
-  regex: new RegExp(/^\|\s*(\d{4,13})\s*\|\s+/),
+  regex: new RegExp(/^\|\s*(\d{4,13})\s*\|/),
   description: 'Create or add to a thread with an ID',
+  example: '|<thread ID>| [message](Optional)',
   category: 'custom threads',
   permissions: 'general',
   execute: async (message) => {
     await message.delete();
```
[the remark this answers is not in the document]
Comment (charlotte-zhuang) [reply src/nonPrefixCommands/threads/threadMessage.js:17]: I think this might be causing other errors too, when you later reference a message that was deleted. I'm not sure though.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 6]
```diff
@@ -1,17 +1,298 @@
...
+      .addField('Topic', `${createThread.responseData.topic}`);
+    return await message.channel.send(threadCreateEmbed);
+  }
+}
+/**
+ * @param {Enumeration} data an enumeration containing all 
+ * necessary data to do APIfunction calls, and make embeds
+ * @param {Discord.Message} message the original message 
+ * that started this command
+ * @param {araay} queryThread the response from FIND_THREAD(data).
```
[the remark this answers is not in the document]
Comment (charlotte-zhuang) [reply src/nonPrefixCommands/threads/threadMessage.js:136]: `queryThread` is an `Object`, not an `Array`.

**Review by charlotte-zhuang — CHANGES_REQUESTED** [review]
See comments:

1. There are a few ways to make your command error
2. The user's original message probably shouldn't be deleted, unless it has no arguments (for viewing a thread)
3. Some doc comments are incorrect
4. A few other items to consider

Also, the `|<thread ID>|` no arguments command should display a full thread (all its messages). It's how users actually use a thread. I don't think this needs to be added right now, but that's the goal.

We also need to change the `Command.js` class to remove the `params` member and add a `disabled` member (like what Sam did for the help command). Disabled non-prefix commands should be ignored by `NonPrefixRegexCreator.js` and should not be added to the command map in `NonPrefixHandler.js`. This might be done in a separate branch.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 7]
```diff
@@ -1,17 +1,298 @@
...
+    message.channel
+      .send('Error adding')
+      .then((msg) => msg.delete(10000));
+    return;
+  }
+  const addedEmbed = new Discord.RichEmbed()
+    .setColor('#301934')
+    .setTitle(`Added message to Thread ${data.threadID}`)
+    .addField('Message', `${data.threadMsg}`);
+  return await
```
[the remark this answers is not in the document]
Comment (charlotte-zhuang) [reply src/nonPrefixCommands/threads/threadMessage.js]: bump

**Comment by jerrylee17** [comment]
[Here](https://github.com/SCE-Development/SCE-discord-bot/compare/patrick) there are 2 files that have been changed

However, this PR says that there are 8 files changed.
image

github hello?????????????????????????????????????

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 8]
```diff
@@ -1,17 +1,315 @@
...
+  };
+  let threadListEmbed = new Array(page.maxPage);
+  threadListEmbed[0] = new Discord.RichEmbed()
+    .setColor('#301934')
+    .setTitle(`All threads that start with ID: ${data.threadID}`);
+  if(messageMode)
+  {
+    threadListEmbed[0].setDescription('Choose one! Example: type "1"');
+  }
+  queryThread.responseData
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:232]: I'm a lazy programmer and you used this functionality more than once so maybe make it a function

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 9]
```diff
@@ -1,17 +1,315 @@
...
+    await addMessageToThread(data, message, queryThread2);
+  };
+  messageCollector.on('collect', collectMessage);
+  //  when collector is turned off, previous messages
+  emojiCollector.on('end', async () => { 
+    messageCollector.stop();
+  });
+}
+
+async function pagination(data, message, queryThread, messageMode){
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:210]: More comments in the code might be better. I spent like 30 mins trying to understand everything going on (my brain also tired so that might be why).

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 10]
```diff
@@ -1,17 +1,315 @@
...
+    });
+  if(page.maxPage)  
+    threadListEmbed[page.currentPage]
+      .setFooter(`Page: ${page.currentPage+1}/${page.maxPage+1}`);
+  let currentMsg = await message.channel
+    .send(threadListEmbed[page.currentPage]);
+  // if more than 1 maxPage then allow mult page
+  if(page.maxPage)
+    await currentMsg.react('⬅️');
+  await currentMsg.react('⏹️');
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:247]: I don't think the stop functionality is really necessary. It is usually helpful to keep the embed up for a while.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 11]
```diff
@@ -1,17 +1,315 @@
...
+    }
+    currentMsg
+      .edit(threadListEmbed[page.currentPage])
+      .catch();
+    await reaction.remove(reaction.users.last().id);
+  };
+    // turn on the collector
+  emojiCollector.on('collect', checkEmoji);
+  emojiCollector.on('end', async () => { 
+    currentMsg.delete();
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:312]: Don't delete - might be helpful to have later

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 12]
```diff
@@ -1,17 +1,315 @@
...
+ **/
+async function multipleThreadResults(data, message, queryThread){
+  let emojiCollector 
+   = await pagination(data, message, queryThread, true);
+  /** 
+   * Create Message collector
+   * @param {Discord.Message} m is message collected by discord
+   * new filter for message collector 
+   **/ 
+  let filter = (m) => {
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:177]: Make this so that if the user has to immediately choose right after the modal. End the messagecollector after collecting 1 message (from the author), and if the message doesn't match, reply with something like "Sorry that's not a valid index blah blah blah".

Also, does the messageCollector collect messages from people that are not the author?

**Review by jerrylee17 — CHANGES_REQUESTED** [review]
1. Enable `commands/threads/thread` and `commands/threads/threadmanager` in this pr since the thread functionality will be done. 
2. The functionality for showing the full thread does not work for me. It shows 1 thread only. See this image:
image
3. When I enter an invalid thread ID, I get this error. Probably display an embed saying "there are no threads that start with blahblahblah"
image
4. I think the current method of creating a new thread is janky. Perhaps we can make the rest of the message the topic (Basically same thing as doing `s!thread`. Also do a check on the topic length and if it's over a certain amount, give it a warning.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 13]
```diff
@@ -1,17 +1,315 @@
...
+      guildID : message.guild.id,
+      threadMsg : messageMode ? 
+        /^\|\s*(\d{4,13})\s*\|\s*(.+)/.exec(message)[2] : '',
+    };
+    //  MESSAGE HANDLING
+    let queryThread = await THREAD_ID_QUERY(data.threadID);
+    if(queryThread.error)
+    {
+      message.channel
+        .send('Error')
```
[the remark this answers is not in the document]
Comment (jerrylee17) [reply src/nonPrefixCommands/threads/threadMessage.js:39]: ^^ Makes it more verbose so people know where the error is

File: src/commands/threads/thread.js [hunk 14]
```diff
@@ -1,15 +1,279 @@
```
[the remark this answers is not in the document]
Comment (jerrylee17) [reply src/commands/threads/thread.js:1]: bump

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 15]
```diff
@@ -1,17 +1,315 @@
...
+    }
   }
 });
+/** 
+ * @param {Object} data an enumeration containing all 
+ * necessary data to do APIfunction calls, and make embeds
+ * @param {Discord.Message} message the original message 
+ * that started this command
+ **/
+async function createNewThread(data, message){
```
[the remark this answers is not in the document]
Comment (jerrylee17) [reply src/nonPrefixCommands/threads/threadMessage.js:103]: Probably should be done in a new PR

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 16]
```diff
@@ -1,17 +1,315 @@
...
+          .then((msg) => msg.delete(10000));
+        return false;
+      });
+  };
+
+  const confirmed = await confirmAction(data.topic);  
+  if(!confirmed)
+  {
+    await message.channel
+      .send('Not created')
```
[the remark this answers is not in the document]
Comment (jerrylee17) [reply src/nonPrefixCommands/threads/threadMessage.js:141]: bump

[the remark this answers is not in the document]
Comment (jerrylee17) [reply src/nonPrefixCommands/threads/threadMessage.js:141]: bump

**Review by jerrylee17 — CHANGES_REQUESTED** [review]
1. If I type an invalid ID, it deletes my message. Probably if theres nothing returned, display an embed telling them that there's nothing returned so the user isn't like "what happened to my message"
Also, I get this error:
image
2. I think the current method of creating a new thread is janky. Perhaps we can make the rest of the message the topic (Basically same thing as doing s!thread. Also do a check on the topic length and if it's over a certain amount, give it a warning.
3. Enable commands/threads/thread and commands/threads/threadmanager in this pr since the thread functionality will be done.

File: src/nonPrefixCommands/nonPrefixFunctions/pagination.js [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+  * the pages will follow
+  * @param {Discord.Message} message 
+  * Message that was sent to start this command
+  * @param {import('../../APIFunctions/thread').Thread} threads 
+  * An array containing ALL items that will be displayed in the "book"
+  * @param {Boolean} messageMode 
+  * @param {Number} ITEM_PER_PAGE 
+  * @return {Discord.ReactionCollector}
+  */
+async function pagination(templateEmbed, message,
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/nonPrefixFunctions/pagination.js:26]: If there's only one page, don't react with arrows and wait for reactions (just display the content).

File: src/nonPrefixCommands/nonPrefixFunctions/pagination.js [hunk 18]
```diff
@@ -0,0 +1,243 @@
...
+  * @param {Discord.RichEmbed} templateEmbed 
+  * template containing Title, Description, and color that
+  * the pages will follow
+  * @param {import('../../APIFunctions/thread').Thread} threads 
+  * An array containing ALL items that will be displayed in the "book"
+  * @param {Page} page
+  * @param {Boolean} messageMode
+  * @return {Discord.RichEmbed}
+  */
+async function createThreadEmbed(threads,
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/nonPrefixFunctions/pagination.js:165]: Error if a header is over 256 characters.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 19]
```diff
@@ -1,17 +1,110 @@
...
+     * False = User looking at threads with starting ID
+     */
+    const createMode = /^\|\s*(\d{4,13})\s*\|\s*(.+)/.test(message);
+    /**
+     * @description
+     * Object containing all necessary data to use
+     * any APIFunction
+     * @type {Data}
+     */
+    const data = {
```
Comment (charlotte-zhuang) [inline src/nonPrefixCommands/threads/threadMessage.js:44]: I think it's much better practice to not make a generic dump of all the variables you need (most functions don't need most of these variables, so don't pass it to them).

**Review by charlotte-zhuang — CHANGES_REQUESTED** [review]
## Changes Made

I made a bunch of changes to the API functions: the arguments are different now and all functions require a guild ID now.
I also added the disabled option to non-prefix commands.

## Changes Reqested

1. Error if a user deletes a message in a thread. This is because you try to display a deleted message. When you encounter a message that has been deleted, use the `DELETE_THREADMESSAGE` function to delete the message from the thread, then skip the message.
2. Error if a user adds a message longer than 256 characters. This is because you display messages as headers, which are limited at 256 characters. Messages should be displayed in the body so that they're not limited (I think).
3. Multi-line messages should be supported. This isn't exactly critical, but definitely preferred.
4. The `nonPrefixCommands` folder should only contain commands; don't put extra functions inside that `nonPrefixFunctions` folder. Right now the functions in that folder are specific to your commands anyway, so move them into the same file. Later on we'll want to add some of those functions to `util` and some of them in a new `components` folder.
5. Your doc comments are much better, but still not formatted correctly. If you aren't sure how to write a doc comment, it's better to change it into a normal comment or keep it simple.

File: src/nonPrefixCommands/nonPrefixFunctions/pagination.js (hunk 17 above)
Comment (lupatrick13) [reply src/nonPrefixCommands/nonPrefixFunctions/pagination.js:26]: I think i've already done that. It checks the max pages in the pagination function

Comment (charlotte-zhuang) [reply src/nonPrefixCommands/nonPrefixFunctions/pagination.js:26]: The code doesn't error, but the reactions still show up at the bottom if that makes sense.

**Comment by charlotte-zhuang** [comment]
## Changes Made

1. So I forgot to push my last commit, this time I actually fixed the API functions.
2. Moved functions into one file (they can be moved into a util or components folder later)
3. Multiple matched threads now display the last message, not the first. Messages that have been deleted on discord are also deleted on the database.
4. Showing a thread now includes the first message. The first message isn't special (the original message creating the thread could have been deleted)
5. Added catching to message deletes.
6. Other bug fixes

Things got toasty while I made all these changes, so another review would be good.

**Review by charlotte-zhuang — COMMENTED** [review]
Made changes.

File: src/APIFunctions/thread.js [hunk 20]
```diff
@@ -3,51 +3,42 @@ const { DISCORD_API_URL } = require('../../config.json');
...
+ * @param {String} args.guildID
+ *
+ * @returns {Promise<ThreadManyPayload>} All threads.
  */
-const THREAD_QUERY = async () => {
+const THREAD_QUERY = async args => {
   const threadQuery = gql`
-    {
-      threadMany {
+    query($guildID: String!) {
```
Comment (jerrylee17) [inline src/APIFunctions/thread.js:40]: Enable this query to take in guildID, channelID optionally.
```
query($guildID: String, $channelID: String) {
  threadMany(filter: { guildID: $guildID, channelID: $channelID }) {
         ...
```

File: src/APIFunctions/thread.js [hunk 21]
```diff
@@ -75,16 +66,26 @@ const THREAD_QUERY = async () => {
...
+ * @param {String} args.guildID
+ * @param {String} args.channelID
  *
- * @returns {ThreadManyPayload} All threads with an ID that start with
+ * @returns {Promise<ThreadManyPayload>} All threads with an ID that start with
  * threadID.
  */
-const THREAD_ID_QUERY = async threadID => {
+const THREAD_ID_QUERY = async args => {
+  args = { ...args };
```
Comment (jerrylee17) [inline src/APIFunctions/thread.js:78]: Aren't you just assigning args to itself here? Is there a purpose for this statement?

File: src/commands/threads/thread.js [hunk 22]
```diff
@@ -19,14 +19,14 @@ module.exports = new Command({
...
-  disabled: true,
+  disabled: false,
   execute: async (message, args) => {
     const param = args.join(' ').trim();
 
     if (param === 'active' || param === 'all') {
       // Show threads
       message.delete(5000).catch(() => null);
-      const response = await THREAD_QUERY();
+      const response = await THREAD_QUERY({ guildID: message.guild.id });
```
Comment (jerrylee17) [inline src/commands/threads/thread.js:29]: Query by both guildID and channelID

File: src/commands/threads/thread.js [hunk 23]
```diff
@@ -43,28 +43,26 @@ module.exports = new Command({
       const fields = [];
       for (let i = 0; i < response.responseData.length; i++) {
         const thread = response.responseData[i];
-        if (
-          thread.guildID !== message.guild.id ||
-          thread.channelID !== message.channel.id
-        ) {
+        if (thread.channelID !== message.channel.id) {
```
Comment (jerrylee17) [inline src/commands/threads/thread.js:46]: This can be omitted if you query by both guildID and channelID

File: src/APIFunctions/thread.js [hunk 24]
```diff
@@ -75,16 +66,26 @@ const THREAD_QUERY = async () => {
...
+ * @param {String} args.threadID
+ * @param {String} args.guildID
+ * @param {String} args.channelID
  *
- * @returns {ThreadManyPayload} All threads with an ID that start with
+ * @returns {Promise<ThreadManyPayload>} All threads with an ID that start with
  * threadID.
  */
-const THREAD_ID_QUERY = async threadID => {
+const THREAD_ID_QUERY = async args => {
```
Comment (jerrylee17) [inline src/APIFunctions/thread.js:77]: `THREAD_ID_QUERY` and `THREAD_QUERY` can be combined into a single query. It may require some fixing up such as checking for `args.threadID` parameter but this would overall make the queries simpler from the client side. This would be easier if implemented in this pull request since the longer we wait the harder it will be to refactor.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 25]
```diff
@@ -1,17 +1,618 @@
-
+const Discord = require('discord.js');
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:1]: This file is 619 lines... We should split it into multiple files to make it more readable.
All those functions after line 119 --> Put it into a new file and import them in.

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 26]
```diff
@@ -1,17 +1,618 @@
...
+          threadQuery,
+          createMode,
+          data.threadMsg
+        );
+    }
+  },
+});
+
+/**
+ * @param {import('../threads/threadMessage').Data} data
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:119]: Interesting docstring... the import is not required since it is in the same file.

**Review by jerrylee17 — CHANGES_REQUESTED** [review]
Review #1: Everything but the 619 line file has been commented.

`src/nonPrefixCommands/threads/threadMessage.js` is pretty tough to review. Probably move the functions to a separate file and import them from here.

**Review by jerrylee17 — CHANGES_REQUESTED** [review]
Review 2: Bug I found

This functionality is actually really developed and I think we are almost done!

1. I unintentionally added a message to 2 threads with 1 message. How did I do this?
```
me: |8994| add this message
bot: pops up a modal of all threads that start with 8994
me: |89944| add this other message
bot: pops up another modal of all threads that start with 89944
me: 2
bot: adds "add this message" to the 2nd result in the first modal
bot: adds "add this other message" to the 2nd result in the second modal
```
Solution: 
Only collect the message sent by the user immediately following the modal. In the example, the bot would make the first modal inactive as soon as the user enters an invalid number.

2. How to get an error creating thread
```
Channel 1: I have a thread with the ID 8994283609061
In Channel 2:
me: |8994283609061| jerry makes a thread with an existing ID
bot: Error creating
```
Suggestion: 
Instead of giving a vague " Error Creating", go with something like "ID has been taken. Choose a different ID or let the bot autogenerate it with `s!thread` blah blah blah".

File: src/APIFunctions/thread.js (hunk 21 above)
Comment (charlotte-zhuang) [reply src/APIFunctions/thread.js:78]: This is necessary to avoid changing the input object.

File: src/APIFunctions/thread.js (hunk 20 above)
Comment (charlotte-zhuang) [reply src/APIFunctions/thread.js:40]: I thought guildID would be mandatory since you shouldn't query outside your own guild?

File: src/nonPrefixCommands/threads/threadMessage.js (hunk 25 above)
Comment (charlotte-zhuang) [reply src/nonPrefixCommands/threads/threadMessage.js:1]: My issue is that the functions are a bit convoluted. I guess there could be a `nonPrefixCommands/threads/threadMessageUtil/` directory for the extra functions?

Comment (charlotte-zhuang) [reply src/nonPrefixCommands/threads/threadMessage.js:1]: Putting any non command directories in `nonPrefixCommands` would require minor changes to the handlers and `nonPrefixRegexCreator.js`. Patrick used to put the util in `nonPrefixCommands/threadFunctions` which works since normally only commands are stored in second level directories.

**Comment by charlotte-zhuang** [comment]
## Changes for dcd885f

1. Thread API changes
2. Asking user to choose a thread now only waits for one message from the user
3. Fixed bug where thread messages would be deleted on other errors
4. Fixed bug where `threadMessage.js` would error when displaying a messages that did start with `|<id>|` or `s!thread`

File: src/nonPrefixCommands/threads/threadMessage.js (hunk 25 above)
Comment (charlotte-zhuang) [reply src/nonPrefixCommands/threads/threadMessage.js:1]: Moved functions to `threadMessageUtil.js`

**Comment by charlotte-zhuang** [comment]
## Changes for 659d6fd

1. Moved utility functions to `threadMessageUtil.js`
2. Cleaned up comments
3. Minor refactor to functions for readability

**Review by jerrylee17 — APPROVED** [review]
Code looks awesome! It's readable and scalable :+1: 

Some braindump...
1. The error is because `8994283609061` already exists in another channel. Would be better to give a more verbose error message.

image

2. A thread with 50+ messages is going to be nasty to scroll through

image

3. I think we will be using pagination a lot in the future. We should probably make it a component initialized with a class (in another pr). 

4. As clear as the instructions for using threads is, we should probably write up some documentation on it with a reference.

This is the biggest PR in SCE-development right now no cap. Great job @lupatrick13 and @charlotte-zhuang!!!

File: src/nonPrefixCommands/threads/threadMessageUtil.js [hunk 27]
```diff
@@ -0,0 +1,479 @@
...
+
+    if (createMode) addMessageToThread(message, threadQuery2.responseData[0]);
+    else {
+      const topic = threads[index].topic;
+      templateEmbed
+        .setTitle('Thread')
+        .setDescription('')
+        .addField('ID', threads[index].threadID, true)
+        .addField('Topic', topic === null ? 'none' : topic, true);
+      pagination(templateEmbed, message, threadQuery2.responseData, true);
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessageUtil.js:230]: Remove the `true`

**Comment by qsaml** [comment]
things I noticed :

- right now, you're allowed to type a random threadID that may or may not exist and it will ask to create the ID, not sure if that is an intentional feature. maybe it's better to just have threads created by threadmanager
- you can't just randomly type 4 numbers of a threadID after it's created and it will add on to a thread (you have to use the same 4 numbers). example : say a threadID is 59596943, typing `|5959|` will start a thread, but you can't add on to it by typing `|9596|`, it starts a new thread
- basically, add to the new thread message embed to type the same 4 numbers that the thread was created with or the full ID
- I feel like the first message of the thread shouldn't be the message creating the thread, as here
image

regarding jerry's comments I agree with them, but holy moly this might be the best thing I've seen to now

**Review by seemavora — COMMENTED** [review]
I gotta say this is an amazing project, looks pretty good so far!!
image
A few suggestions I have:
1. Maybe at the beginning of your PR include what changes you made and what commands/things you want ur PR reviewer to test, since that was slightly unclear. 
2.  I think this was mentioned before but changing the thread id creates a new thread
3. maybe might also be cool to create names for the threads instead of thread ids its too painful to type in a long number

**Review by seemavora — COMMENTED** [review]
Also i gotta say ur code is CLEANNNN  compared to a lot of PRS i have seen go you!

**Comment by charlotte-zhuang** [comment]
## Changes in 5ab3d4d and 0bea8cd

1. Fixed bug when paginating a thread after selecting from multiple options
2. Added thread ID decoration for readability
3. Moved thread ID functions to util

## Not changed

`threadmessages.js` still creates a new thread when the specified ID is not found. Is there a good reason to keep this?

**Review by charlotte-zhuang — APPROVED** [review]

File: src/commands/threads/thread.js [hunk 28]
```diff
@@ -16,21 +18,20 @@ module.exports = new Command({
   name: 'thread',
   description: 'View active threads or start a new one',
   aliases: [],
-  example: 's!thread <[all | active | none] | topic>',
+  example: `${prefix}thread <all | active | none | topic>`,
```
Comment (jerrylee17) [inline src/commands/threads/thread.js:21]: I like this! We should do this for more commands in a future pr

File: src/nonPrefixCommands/threads/threadMessage.js [hunk 29]
```diff
@@ -1,17 +1,90 @@
...
-      + message.content
-      + '"'
-    );
-  }
+  execute: async message => {
+    const input = /^\[([\d\s-]{4,20})\]\s*(.*)/.exec(message);
+    const threadID = undecorateId(input[1]).substr(0, 13);
+    const body = input[2].length === 0 ? null : input[2];
+
+    if (threadID.length < 4) {
```
Comment (jerrylee17) [inline src/nonPrefixCommands/threads/threadMessage.js:25]: Won't ever happen because of the regex.................................


just kidding nice catch

image

**Review by jerrylee17 — CHANGES_REQUESTED** [review]
1. `npm run lint`
2. Using \` will also add in the extra line breaks and spaces. Use `'` instead
image

File: src/util/threadMessage.js [hunk 30]
```diff
@@ -0,0 +1,477 @@
...
+    creatorID: message.member.id,
+    guildID: message.guild.id,
+    channelID: message.channel.id,
+    topic: topic,
+    messageID: message.id,
+  });
+  if (createThread.error) {
+    message.channel
+      .send(
+        `Error creating. Possibly an existing thread uses ID
```
Comment (jerrylee17) [inline src/util/threadMessage.js:104]: I used
```
'Error creating. Possibly an existing thread uses ID ' +
`${decorateId(threadID)}.`
```

**Comment by charlotte-zhuang** [comment]
## Changes in 31ac029

1. I fixed the string literal formatting issues, and also change that error message since I wasn't sure if it was accurate.
2. Fixed bug where prefixed commands can create topics that are too long.
3. Simplified API function code (sorry, I hate my code)
